Compiler's error reporting does not handle tabs properly

The column pointer will fall out of alignment with what it's supposed to be pointing to if the cli environment differs from scalac's expectations in one of two ways:
1: Tab width is not equivalent to 4 spaces
2: Something is prefixed to scalac's output for which prefix.length is not a multiple of 4
For instance,
```
[ant:scalac] /home/mako/programming/ ... /Code.scala:289: error: type mismatch;
[ant:scalac]  found   : Double
[ant:scalac]  required: Float
[ant:scalac] 				if(dt < 1) f(Math.pow(2, 10 * (dt - 1) - 1))
[ant:scalac]                                                      ^
```
In my terminals, comes out looking roughly like
```
[ant:scalac]                         if(dt < 1) f(Math.pow(2, 10 * (dt - 1) - 1))
[ant:scalac]                                                                ^
```
making it much harder to tell where the error is.

Easy fix, I think. For an error being reported against line:Int, column:Int, scalac's output for the pointer line should be 
```scala
codeAtLine(line).slice(0,column).map{ case '\t' => '\t'; case c:Char => ' ' }+'^'
```
This will always put the ^ in the right place.

--- a/src/reflect/scala/reflect/internal/util/Position.scala
+++ b/src/reflect/scala/reflect/internal/util/Position.scala
@@ -177,7 +177,16 @@ private[util] trait InternalPositionImpl {
   def line: Int           = if (hasSource) source.offsetToLine(point) + 1 else 0
   def column: Int         = if (hasSource) calculateColumn() else 0
   def lineContent: String = if (hasSource) source.lineToString(line - 1) else ""
-  def lineCaret: String   = if (hasSource) " " * (column - 1) + "^" else ""
+  def lineCaret: String   = if (!hasSource) "" else {
+    val buf = new StringBuilder
+    var idx = source.lineToOffset(source.offsetToLine(point))
+    while (idx < point) {
+      buf.append(if (source.content(idx) == '\t') '\t' else ' ')
+      idx += 1
+    }
+    buf.append('^')
+    buf.toString
+  }
   @deprecated("use `lineCaret`", since="2.11.0")
   def lineCarat: String   = lineCaret
 

--- a/src/reflect/scala/reflect/internal/util/Position.scala
+++ b/src/reflect/scala/reflect/internal/util/Position.scala
@@ -34,7 +34,7 @@ class Position extends scala.reflect.api.Position with InternalPositionImpl with
 }
 
 object Position {
-  val tabInc = 8
+  final val tabInc = 8
 
   private def validate[T <: Position](pos: T): T = {
     if (pos.isRange)
@@ -174,12 +174,22 @@ private[util] trait InternalPositionImpl {
   // necessary condition to establish that there is overlap.
   def overlaps(pos: Position): Boolean         = bothRanges(pos) && start < pos.end && pos.start < end
 
-  def line: Int           = if (hasSource) source.offsetToLine(point) + 1 else 0
-  def column: Int         = if (hasSource) calculateColumn() else 0
-  def lineContent: String = if (hasSource) source.lineToString(line - 1) else ""
+  private def line0       = source.offsetToLine(point)
+  private def lineOffset  = source.lineToOffset(line0)
+  def line: Int           = if (hasSource) line0 + 1 else 0
+  def column: Int         = if (!hasSource) 0 else {
+    var idx = lineOffset
+    var col = 0
+    while (idx != point) {
+      col += (if (source.content(idx) == '\t') Position.tabInc - col % Position.tabInc else 1)
+      idx += 1
+    }
+    col + 1
+  }
+  def lineContent: String = if (hasSource) source.lineToString(line0) else ""
   def lineCaret: String   = if (!hasSource) "" else {
     val buf = new StringBuilder
-    var idx = source.lineToOffset(source.offsetToLine(point))
+    var idx = lineOffset
     while (idx < point) {
       buf.append(if (source.content(idx) == '\t') '\t' else ' ')
       idx += 1
@@ -200,16 +210,11 @@ private[util] trait InternalPositionImpl {
         sb.toString
       } else s
     }
-    def errorAt(p: Pos) = {
-      def where     = p.line
-      def content   = escaped(p.lineContent)
-      def indicator = p.lineCaret
-      f"$where: $msg%n$content%n$indicator"
-    }
+    import java.lang.System.{lineSeparator => NL}
     finalPosition match {
       case FakePos(fmsg) => s"$fmsg $msg"
       case NoPosition    => msg
-      case pos           => errorAt(pos)
+      case pos           => s"${pos.line}: ${msg}${NL}${escaped(pos.lineContent)}${NL}${pos.lineCaret}"
     }
   }
   def showDebug: String = toString
@@ -227,16 +232,6 @@ private[util] trait InternalPositionImpl {
   private def asOffset(point: Int): Position = Position.offset(source, point)
   private def copyRange(source: SourceFile = source, start: Int = start, point: Int = point, end: Int = end): Position =
     Position.range(source, start, point, end)
-
-  private def calculateColumn(): Int = {
-    var idx = source.lineToOffset(source.offsetToLine(point))
-    var col = 0
-    while (idx != point) {
-      col += (if (source.content(idx) == '\t') Position.tabInc - col % Position.tabInc else 1)
-      idx += 1
-    }
-    col + 1
-  }
   private def hasSource                      = source ne NoSourceFile
   private def bothRanges(that: Position)     = isRange && that.isRange
   private def bothDefined(that: Position)    = isDefined && that.isDefined

--- a/test/files/neg/implicit-shadow.check
+++ b/test/files/neg/implicit-shadow.check
@@ -4,8 +4,8 @@ it is imported twice in the same scope by
 import C._
 and import B._
 	1.isEmpty
-        ^
+	^
 implicit-shadow.scala:6: error: value isEmpty is not a member of Int
 	1.isEmpty
-          ^
+	  ^
 1 error

--- a/test/files/neg/names-package-method-conflict.check
+++ b/test/files/neg/names-package-method-conflict.check
@@ -1,4 +1,4 @@
 names-package-method-conflict.scala:7: error: There is name conflict between the foo.bar and the package foo.bar.
 	package bar{
-                ^
+	        ^
 1 error

--- a/test/files/neg/t2066b.check
+++ b/test/files/neg/t2066b.check
@@ -3,5 +3,5 @@ def f[T[_]](x: T[Int]): T[Any] (defined in trait A);
  found   : [T(in method f)(in method f)[+_]](x: T(in method f)(in method f)[Int]): T(in method f)(in method f)[Any]
  required: [T(in method f)(in method f)[_]](x: T(in method f)(in method f)[Int]): T(in method f)(in method f)[Any]
 	 def f[T[+_]](x : T[Int]) : T[Any] = x
-             ^
+	     ^
 1 error

--- a/test/files/neg/t2208.check
+++ b/test/files/neg/t2208.check
@@ -1,4 +1,4 @@
 t2208.scala:7: error: type arguments [Any] do not conform to type Alias's type parameter bounds [X <: Test.A]
 	class C extends Alias[Any]   // not ok, normalisation should check bounds before expanding Alias
-                        ^
+	                ^
 1 error

--- a/test/files/neg/t2405.check
+++ b/test/files/neg/t2405.check
@@ -1,8 +1,8 @@
 t2405.scala:8: error: could not find implicit value for parameter e: Int
 		implicitly[Int]
-                          ^
+		          ^
 t2405.scala:6: warning: imported `y` is permanently hidden by definition of method y
 		import A.{x => y}
-                          ^
+		          ^
 1 warning
 1 error

--- a/test/files/neg/t4460b.check
+++ b/test/files/neg/t4460b.check
@@ -1,4 +1,4 @@
 t4460b.scala:7: error: constructor invokes itself
 	  def this() = this() // was binding to Predef.<init> !!
-                       ^
+	               ^
 1 error

--- a/test/files/neg/t4818.check
+++ b/test/files/neg/t4818.check
@@ -2,5 +2,5 @@ t4818.scala:4: error: type mismatch;
  found   : Int(5)
  required: Nothing
 	def f(x: Any) = x match { case Fn(f) => f(5) }
-                                                  ^
+	                                          ^
 1 error

--- a/test/files/neg/t6214.check
+++ b/test/files/neg/t6214.check
@@ -1,4 +1,4 @@
 t6214.scala:5: error: missing parameter type
   	m { s => case class Foo() }
-            ^
+  	    ^
 1 error

--- a/test/files/neg/t6258.check
+++ b/test/files/neg/t6258.check
@@ -2,12 +2,12 @@ t6258.scala:2: error: missing parameter type for expanded function
 The argument types of an anonymous function must be fully known. (SLS 8.5)
 Expected type was: PartialFunction[?, Int]
 	val f : PartialFunction[_, Int] = { case a : Int => a } // undefined param
-                                          ^
+	                                  ^
 t6258.scala:5: error: missing parameter type for expanded function
 The argument types of an anonymous function must be fully known. (SLS 8.5)
 Expected type was: PartialFunction[?,Int]
 	foo { case a : Int => a } // undefined param
-            ^
+	    ^
 t6258.scala:22: error: missing parameter type for expanded function
 The argument types of an anonymous function must be fully known. (SLS 8.5)
 Expected type was: PartialFunction[?,Any]

--- a/test/files/neg/t6558.check
+++ b/test/files/neg/t6558.check
@@ -6,5 +6,5 @@ t6558.scala:7: error: not found: type typeparam
            ^
 t6558.scala:10: error: not found: type valueparam
   	@valueparam x: Any
-         ^
+  	 ^
 3 errors

--- a/test/files/neg/t9438.check
+++ b/test/files/neg/t9438.check
@@ -1,0 +1,11 @@
+t9438.scala:3: error: type mismatch;
+ found   : Double(3.14)
+ required: Int
+	val i: Int = 3.14
+	             ^
+t9438.scala:5: error: type mismatch;
+ found   : Double(3.14)
+ required: Int
+		3.14
+		^
+2 errors

--- a/test/files/neg/t9438.scala
+++ b/test/files/neg/t9438.scala
@@ -1,0 +1,6 @@
+
+class C {
+	val i: Int = 3.14
+	def f(): Int =
+		3.14
+}

